### ISSUE TITLE
With commandTimeout option set, active timers are not cleared on client disconnection

### ISSUE THREAD

**Issue by gregplaysguitar** [luin/ioredis#1535]
certainly appears that ioredis doesn't clean up its timers when the client is disconnected :face_with_raised_eyebrow: 

With commandTimeout set, a timer is set up here:
https://github.com/luin/ioredis/blob/eabb70ed835828ad85b602a217992486e05d2218/lib/Command.ts#L292-L300

the only place this timer gets cleared is here

https://github.com/luin/ioredis/blob/eabb70ed835828ad85b602a217992486e05d2218/lib/Command.ts#L354-L355

and _that_ only gets called on successful resolution of the command...

https://github.com/luin/ioredis/blob/eabb70ed835828ad85b602a217992486e05d2218/lib/Command.ts#L313

The result is that if the client is disconnected while there is an active command timeout, there ends up being an orphaned timer in the nodejs process. In my case, this is causing tests and lambda executions to hang until the timer clears.

Possible simple fix is just to clear the timer on promise rejection too, here? I'm not familiar enough with the codebase to know if this is right though. If a maintainer could confirm this is a good idea, I can open a PR to implement.

 https://github.com/luin/ioredis/blob/eabb70ed835828ad85b602a217992486e05d2218/lib/Command.ts#L316

**Comment by luin** [comment]
We can't clear the timeout on rejections because a rejection doesn't necessarily mean a command is failed, instead, the connection can be reconnected and the command can be resent so the idea of `commandTimeout` is to avoid the command hanging forever waiting for the reconnection.

What's in my mind is to track the timers with a list and clear them when disconnected. Not sure it's worth though.

**Comment by gregplaysguitar** [comment]
Fair enough - fwiw we are moving away from using Redis in lambdas, so it won't be a problem for us much longer. I do think it's a genuine problem but understand if the effort to fix isn't worthwhile from your point of view

**Comment by luin** [comment]
Yeah I think we can probably clear the timeouts on "flushCommandQueue" which will be called on the "end" event.

**Comment by akeating** [comment]
tbh I'm struggling to understand the design of this part of the code. E.g. if I terminate retries via a custom `retryStrategy`, any queued commands will be rejected with `Connection is closed`. The client status is `end`. This makes sense to me.

But the `setTimeout` set in the `sendCommand` method is never cleared, even though the promise was rejected.

It seems to me that this `setTimeout` should be cleared on rejection too?

> We can't clear the timeout on rejections because a rejection doesn't necessarily mean a command is failed

What are the use cases for actioning a command after it has been rejected? Let's take this as an example:
```
await redis.get("foo");
...
```

Here, the promise gets rejected and my code does whatever handling makes sense for my application.
What I'm hearing you say is that we need to keep the `setTimeout` around because the command will be actioned later?

**Comment by afonsojanu** [comment]
Went digging on this since it looked like it might still be open. Checked `_clearTimers()` (Command.ts) and it turns out it's already wired into the `reject` path, so `flushQueue` clearing the command queue on disconnect does clear the timeout timer along with it.

Wrote a quick repro with sinon fake timers: issue a command against a server that never replies with commandTimeout set high, disconnect while it's pending, wait for the actual `end` event, then check the timer count. It goes back to 0 every time. There's a brief window where a reconnect timeout gets scheduled and then cancelled once manuallyClosing is checked, which might be what looked like a leak depending on when you were sampling, but nothing sticks around once the client settles into `end`.

Looks like this got fixed as a side effect of the Command/Redis rewrite at some point rather than a deliberate patch for this issue, so it never got closed out. Might be worth double checking against whatever version you originally saw the leak on, but on current main I can't reproduce it anymore.
